File: CTF_Writeups/scripts_python/NovruzCTF_Speedrun.py
import re
import socket
import sys
import time

RULE_RE = re.compile(rb"'(.)'\s*=>\s*(\w+)")

OP_FALLBACK = {
    ord('+'): b'ADD',
    ord('-'): b'SUB',
    ord('*'): b'MUL',
    ord('/') : b'DIV',
    ord('%'): b'MOD',
    ord('^'): b'XOR',
    ord('&'): b'AND',
    ord('|'): b'OR',
}


def python_div(a, b):
    if b == 0:
        return 0
    return a // b


def python_mod(a, b):
    if b == 0:
        return 0
    return a % b


def int_pow(a, b):
    if b < 0:
        return 0
    return pow(a, b)


def apply_op(sym, a, b, op_map):
    op = op_map.get(sym) or OP_FALLBACK.get(sym)
    if op is None:
        return 0
    if op == b'ADD':
        return a + b
    if op == b'SUB':
        return a - b
    if op == b'MUL':
        return a * b
    if op == b'DIV':
        return python_div(a, b)
    if op == b'MOD':
        return python_mod(a, b)
    if op == b'XOR':
        return a ^ b
    if op == b'AND':
        return a & b
    if op == b'OR':
        return a | b
    if op == b'LSHIFT':
        return 0 if b < 0 or b > 63 else (a << b)
    if op == b'RSHIFT':
        return 0 if b < 0 or b > 63 else (a >> b)
    if op == b'POW':
        return int_pow(a, b)
    return 0


def parse_expr(expr, i, op_map):
    n = len(expr)
    while i < n and expr[i] == 32:
        i += 1
    if i >= n:
        return 0, i

    if expr[i] == ord('('):
        i += 1
        left, i = parse_expr(expr, i, op_map)
        while i < n and expr[i] == 32:
            i += 1
        if i < n and expr[i] == ord(')'):
            return left, i + 1
        op = expr[i]
        i += 1
        right, i = parse_expr(expr, i, op_map)
        while i < n and expr[i] == 32:
            i += 1
        if i < n and expr[i] == ord(')'):
            i += 1
        return apply_op(op, left, right, op_map), i

    start = i
    if i < n and expr[i] in (ord('+'), ord('-')):
        i += 1
    while i < n and 48 <= expr[i] <= 57:
        i += 1
    if start == i:
        return 0, i
    return int(expr[start:i]), i


def evaluate(expr, op_map):
    v, _ = parse_expr(expr, 0, op_map)
    return v


def main():
    addr = ("142.93.12.237", 1337)
    if len(sys.argv) > 1:
        host, port = sys.argv[1].split(':')
        addr = (host, int(port))

    s = socket.create_connection(addr, timeout=10)
    s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)

    op_map = {}
    solved = 0
    start = time.time()

    buf = b""
    while True:
        chunk = s.recv(65536)
        if not chunk:
            break
        buf += chunk
        while True:
            nl = buf.find(b'\n')
            if nl < 0:
                break
            line = buf[:nl]
            buf = buf[nl + 1:]
            if not line:
                continue

            if b'RULES:' in line:
                op_map = {m.group(1)[0]: m.group(2).upper() for m in RULE_RE.finditer(line)}
                continue

            if b'Calculate: ' in line:
                expr = line.split(b'Calculate: ', 1)[1].strip()
                res = evaluate(expr, op_map)
                s.sendall(str(res).encode() + b'\n')
                solved += 1
                if solved % 64 == 0:
                    elapsed = time.time() - start
                    sys.stderr.write(f"[*] {solved}/256 at {elapsed:.2f}s\n")
                continue

            trimmed = line.strip()
            if trimmed and not trimmed.startswith(b'####') and b'PROTOCOL UPDATE' not in trimmed:
                print(trimmed.decode(errors='ignore'))

    elapsed = time.time() - start
    print(f"[*] Solved: {solved}/256 in {elapsed:.2f}s")


if __name__ == '__main__':
    main()

File: scripts_go/NovruzCTF_Speedrun.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"net"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var opMap map[byte]string
var rulesRe = regexp.MustCompile(`'(.)'\s*=>\s*(\w+)`)

var rulesTag = []byte("RULES:")
var calcTag = []byte("Calculate: ")
var protoTag = []byte("PROTOCOL UPDATE")

func main() {
	addr := "142.93.12.237:1337"
	if len(os.Args) > 1 {
		addr = os.Args[1]
	}

	fmt.Printf("[*] Connecting to %s...\n", addr)
	start := time.Now()

	conn, err := net.DialTimeout("tcp", addr, 10*time.Second)
	if err != nil {
		fmt.Println("[-] Failed:", err)
		os.Exit(1)
	}
	defer conn.Close()

	if tc, ok := conn.(*net.TCPConn); ok {
		_ = tc.SetNoDelay(true)
		_ = tc.SetReadBuffer(1 << 20)
		_ = tc.SetWriteBuffer(1 << 20)
		_ = tc.SetKeepAlive(true)
		_ = tc.SetKeepAlivePeriod(5 * time.Second)
	}

	fmt.Printf("[*] Connected in %v\n", time.Since(start))

	opMap = make(map[byte]string)
	solved := 0
	answerBuf := make([]byte, 0, 8192)

	reader := bufio.NewReaderSize(conn, 1<<20)
	for {
		line, err := reader.ReadBytes('\n')
		if len(line) == 0 && err != nil {
			if err == io.EOF {
				break
			}
			fmt.Fprintln(os.Stderr, "read error:", err)
			break
		}

		line = bytes.TrimSpace(line)
		if len(line) == 0 {
			if err != nil && err != io.EOF {
				break
			}
			continue
		}

		if bytes.Contains(line, rulesTag) {
			parseRules(string(line))
			continue
		}

		if calcIdx := bytes.Index(line, calcTag); calcIdx >= 0 {
			expr := bytes.TrimSpace(line[calcIdx+len(calcTag):])
			result := evaluate(string(expr))
			answerBuf = strconv.AppendInt(answerBuf[:0], result, 10)
			answerBuf = append(answerBuf, '\n')
			_, _ = conn.Write(answerBuf)
			solved++
			if solved%64 == 0 {
				fmt.Fprintf(os.Stderr, "[*] %d/256 at %v\n", solved, time.Since(start))
			}
			continue
		}

		if len(line) > 0 && !bytes.HasPrefix(line, []byte("####")) && !bytes.Contains(line, protoTag) {
			fmt.Println(string(line))
		}

		if err == io.EOF {
			break
		}
	}

	elapsed := time.Since(start)
	fmt.Printf("\n[*] Solved: %d/256 in %v (%.1fms/round)\n", solved, elapsed, float64(elapsed.Milliseconds())/float64(max(solved, 1)))
}

func parseRules(line string) {
	opMap = make(map[byte]string)
	for _, m := range rulesRe.FindAllStringSubmatch(line, -1) {
		opMap[m[1][0]] = strings.ToUpper(m[2])
	}
}

func evaluate(expr string) int64 {
	pos := 0
	return parseExpr(expr, &pos)
}

func parseExpr(expr string, pos *int) int64 {
	for *pos < len(expr) && expr[*pos] == ' ' {
		*pos++
	}
	if *pos >= len(expr) {
		return 0
	}

	if expr[*pos] == '(' {
		*pos++
		left := parseExpr(expr, pos)
		for *pos < len(expr) && expr[*pos] == ' ' {
			*pos++
		}
		if *pos < len(expr) && expr[*pos] == ')' {
			*pos++
			return left
		}
		op := expr[*pos]
		*pos++
		right := parseExpr(expr, pos)
		for *pos < len(expr) && expr[*pos] == ' ' {
			*pos++
		}
		if *pos < len(expr) && expr[*pos] == ')' {
			*pos++
		}
		return applyOp(op, left, right)
	}

	start := *pos
	if *pos < len(expr) && (expr[*pos] == '-' || expr[*pos] == '+') {
		*pos++
	}
	for *pos < len(expr) && expr[*pos] >= '0' && expr[*pos] <= '9' {
		*pos++
	}
	if start == *pos {
		return 0
	}
	n, _ := strconv.ParseInt(expr[start:*pos], 10, 64)
	return n
}

func applyOp(sym byte, a, b int64) int64 {
	opName, ok := opMap[sym]
	if !ok {
		switch sym {
		case '+':
			return a + b
		case '-':
			return a - b
		case '*':
			return a * b
		case '/':
			if b == 0 {
				return 0
			}
			return a / b
		case '%':
			if b == 0 {
				return 0
			}
			return pythonMod(a, b)
		case '^':
			return a ^ b
		case '&':
			return a & b
		case '|':
			return a | b
		}
		return 0
	}

	switch opName {
	case "ADD":
		return a + b
	case "SUB":
		return a - b
	case "MUL":
		return a * b
	case "DIV":
		if b == 0 {
			return 0
		}
		return pythonDiv(a, b)
	case "MOD":
		if b == 0 {
			return 0
		}
		return pythonMod(a, b)
	case "XOR":
		return a ^ b
	case "AND":
		return a & b
	case "OR":
		return a | b
	case "LSHIFT":
		if b < 0 || b > 63 {
			return 0
		}
		return a << uint(b)
	case "RSHIFT":
		if b < 0 || b > 63 {
			return 0
		}
		return a >> uint(b)
	case "POW":
		return intPow(a, b)
	}
	return 0
}

func pythonMod(a, b int64) int64 {
	r := a % b
	if r != 0 && (r < 0) != (b < 0) {
		r += b
	}
	return r
}

func pythonDiv(a, b int64) int64 {
	r := a / b
	if (a^b) < 0 && r*b != a {
		r--
	}
	return r
}

func intPow(base, exp int64) int64 {
	if exp < 0 {
		return 0
	}
	result := int64(1)
	for exp > 0 {
		if exp%2 == 1 {
			result *= base
		}
		exp /= 2
		base *= base
	}
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
